### PR TITLE
Fix a race in the scheduler.

### PR DESCRIPTION
**Pull request by mikedanese**
Loop over priorityConfigs seperately. The node loop can only safely
modify result[i][index]. Before this change it sometimes modified
result[i] concurrently with other loops.

Verified with `go test -race -run=TestGenericScheduler/test_6 -test.v -count 1000 ./pkg/scheduler/core`

```
==================== Test output for //pkg/scheduler/core:go_default_test:
==================
WARNING: DATA RACE
Read at 0x00c0005e8ed0 by goroutine 22:
  k8s.io/kubernetes/pkg/scheduler/core.PrioritizeNodes.func2()
      pkg/scheduler/core/generic_scheduler.go:667 +0x2ea
  k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.ParallelizeUntil.func1()
      staging/src/k8s.io/client-go/util/workqueue/parallelizer.go:65 +0x9e

Previous write at 0x00c0005e8ed0 by goroutine 21:
  k8s.io/kubernetes/pkg/scheduler/core.PrioritizeNodes.func2()
      pkg/scheduler/core/generic_scheduler.go:668 +0x450
  k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.ParallelizeUntil.func1()
      staging/src/k8s.io/client-go/util/workqueue/parallelizer.go:65 +0x9e

Goroutine 22 (running) created at:
  k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.ParallelizeUntil()
      staging/src/k8s.io/client-go/util/workqueue/parallelizer.go:57 +0x1a3
  k8s.io/kubernetes/pkg/scheduler/core.PrioritizeNodes()
      pkg/scheduler/core/generic_scheduler.go:682 +0x592
  k8s.io/kubernetes/pkg/scheduler/core.(*genericScheduler).Schedule()
      pkg/scheduler/core/generic_scheduler.go:186 +0x77d
  k8s.io/kubernetes/pkg/scheduler/core.TestGenericScheduler.func1()
      pkg/scheduler/core/generic_scheduler_test.go:464 +0x91f
  testing.tRunner()
      GOROOT/src/testing/testing.go:827 +0x162

Goroutine 21 (running) created at:
  k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.ParallelizeUntil()
      staging/src/k8s.io/client-go/util/workqueue/parallelizer.go:57 +0x1a3
  k8s.io/kubernetes/pkg/scheduler/core.PrioritizeNodes()
      pkg/scheduler/core/generic_scheduler.go:682 +0x592
  k8s.io/kubernetes/pkg/scheduler/core.(*genericScheduler).Schedule()
      pkg/scheduler/core/generic_scheduler.go:186 +0x77d
  k8s.io/kubernetes/pkg/scheduler/core.TestGenericScheduler.func1()
      pkg/scheduler/core/generic_scheduler_test.go:464 +0x91f
  testing.tRunner()
      GOROOT/src/testing/testing.go:827 +0x162
==================
--- FAIL: TestGenericScheduler (0.01s)
    --- FAIL: TestGenericScheduler/test_6 (0.00s)
        testing.go:771: race detected during execution of test
    testing.go:771: race detected during execution of test
FAIL
```

/kind bug
/sig scheduling

```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/core/generic_scheduler.go [hunk 1]
```diff
@@ -657,29 +657,37 @@ func PrioritizeNodes(
...
 	}
 
-	processNode := func(index int) {
-		nodeInfo := nodeNameToInfo[nodes[index].Name]
+	// DEPRECATED: we can remove this when all priorityConfigs implement the
+	// Map-Reduce pattern.
+	workqueue.ParallelizeUntil(context.TODO(), 16, len(priorityConfigs), func(i int) {
+		priorityConfig := priorityConfigs[i]
+		if priorityConfig.Function == nil {
+			return
```
Comment (Huang-Wei) [inline pkg/scheduler/core/generic_scheduler.go]: One nit: we can remove L656~L658, and move L657 here:

```suggestion
			results[i] = make(schedulerapi.HostPriorityList, len(nodes))
			return
```

Comment (mikedanese) [reply pkg/scheduler/core/generic_scheduler.go]: Done.

**Review by Huang-Wei — APPROVED** [review]
/lgtm

Thanks @mikedanese for addressing this!

File: pkg/scheduler/core/generic_scheduler.go (hunk 1 above)
Comment (bsalamat) [reply pkg/scheduler/core/generic_scheduler.go]: NVM, this ParallelizeUntil runs for each priorityConfig, not for each node. Confused it with the next one.

Comment (Huang-Wei) [reply pkg/scheduler/core/generic_scheduler.go]: But for a particular `i` (the `i` here is for each `priorityConfig`, not `node`), it only runs once, isn't it?

Unlike the second `ParallelizeUntil` below, it's just iterating `priorityConfigs`.

**Review by bsalamat — APPROVED** [review]
/lgtm
/approve

Thanks, @mikedanese!

File: pkg/scheduler/core/generic_scheduler.go (hunk 1 above)
Comment (Huang-Wei) [reply pkg/scheduler/core/generic_scheduler.go]: Yeap, we commented exactly at the same time :)

**Pull request merged**